File: python-algorithm/ds_1_stack_queue/L155.py
# -*- coding: utf-8 -*-

# 这里是为了用list仿c++等语言的stack功能，所以不要使用python list的其它特性

class MinStack:
    def __init__(self):
        """
        initialize your data structure here.
        """
        self.stk = []
        self.min_stk = []

    def push(self, val: int) -> None:
        self.stk.append(val)
        if not self.min_stk:
            self.min_stk.append(val)
        else:
            if val >= self.min_stk[-1]:
                self.min_stk.append(self.min_stk[-1])
            else:
                self.min_stk.append(val)

    def pop(self) -> None:
        self.min_stk.pop()
        return self.stk.pop()

    def top(self) -> int:
        return self.stk[-1]

    def getMin(self) -> int:
        return self.min_stk[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: python-algorithm/ds_1_stack_queue/L155.cpp

#include <iostream>
#include <stack>

using namespace std;

class MinStack {
public:
    /** initialize your data structure here. */
    stack<int> s1;
    stack<int> s2;
    MinStack() {}

    void push(int x) {
        s1.push(x);
        if(s2.empty())
            s2.push(x);
        else{
            if(x < s2.top())
                s2.push(x);
            else
                s2.push(s2.top());
        }
    }

    void pop() {
        if(!s1.empty()){
            s1.pop();
            s2.pop();
        }
    }

    int top() {
        if(!s1.empty())
            return s1.top();
        throw ;
    }

    int getMin() {
        if(!s2.empty())
            return s2.top();
        throw ;
    }

};


//下面是节省内存的写法，即辅助栈不用存每步对应的最小值

class MinStack2 {
public:
    /** initialize your data structure here. */
    stack<int> s1;
    stack<int> s2;
    MinStack() {

    }

    void push(int x) {
        s1.push(x);
        if(s2.empty() || x <= s2.top() )
            s2.push(x);
    }

    void pop() {
        if(!s1.empty()){
            if(s1.top() == s2.top())
                s2.pop();
            s1.pop();
        }

    }

    int top() {
        if(!s1.empty())
            return s1.top();
        throw ;
    }

    int getMin() {
        if(!s2.empty())
            return s2.top();
        throw ;
    }
};


/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(x);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

int main(){
    return 0;
}